### PR TITLE
Fix TalkNet and HiFiGAN and related files

### PR DESCRIPTION
**Pull request by Oktai15**
* Updated HiFiGAN training blocks and config (main version doesn't work)
* Updated TalkNet and config (main version doesn't work)
* Small changes in audio_to_text.py (load_audio flag)
* Small changes in vocabs.Phonemes/vocabs.Chars (add_blank_at flag)

### CODE REVIEW

File: nemo/collections/asr/data/vocabs.py [hunk 1]
```diff
@@ -80,7 +94,13 @@ def decode(self, tokens: List[int]) -> str:
 class Chars(Base):
     """Chars vocabulary."""
 
-    def __init__(self, punct=True, spaces=False, apostrophe=True):
+    def __init__(
+        self,
+        punct=True,
+        spaces=False,
+        apostrophe=True,
+        add_blank_at="last_but_one",
```
Comment (stasbel) [inline nemo/collections/asr/data/vocabs.py:102]: what's that argument for?

File: examples/tts/conf/talknet-pitch.yaml [hunk 2]
```diff
@@ -55,11 +60,12 @@ model:
         punct: true
         spaces: true
         stresses: false
+        add_blank_at: last
     dataloader_params:
       drop_last: false
       shuffle: false
       batch_size: 64
-      num_workers: 8
+      num_workers: 1
```
Comment (stasbel) [inline examples/tts/conf/talknet-pitch.yaml:68]: having single worker doesn't really make sense, does it?
num_workers=0 should work at least as fast

File: examples/tts/conf/talknet-pitch.yaml [hunk 3]
```diff
@@ -11,6 +11,10 @@ model:
   dropout: 0.1
   separable: true
 
+  # F0 stats (depends on dataset, current for LJSpeech)
+  f0_mean: 178.7767791748047
```
Comment (stasbel) [inline examples/tts/conf/talknet-pitch.yaml:15]: we can precalculate it for any dataset in data layer init, then save it as part of checkpoint (like torch bufffer)

**Comment by stasbel** [comment]
Can you also update latest TalkNet weights in pretrained link?

File: examples/tts/conf/talknet-pitch.yaml (hunk 3 above)
Comment (Oktai15) [reply examples/tts/conf/talknet-pitch.yaml:15]: @stasbel it is good idea for small dataset, but for large dataset it can be impractical.

File: examples/tts/conf/talknet-pitch.yaml (hunk 2 above)
Comment (Oktai15) [reply examples/tts/conf/talknet-pitch.yaml:68]: In my experiments, num_workers=1 works slightly faster

File: nemo/collections/asr/data/vocabs.py (hunk 1 above)
Comment (Oktai15) [reply nemo/collections/asr/data/vocabs.py:102]: For using ASR models in NeMo we assume that blank index in vocab is last. For saving back compatibility with previous version of `vocabs.py` I created this argument.

File: examples/tts/conf/hifigan/hifigan.yaml [hunk 4]
```diff
@@ -36,23 +36,31 @@ model:
...
   num_nodes: 1
   accelerator: ddp
   accumulate_grad_batches: 1
   checkpoint_callback: False  # Provided by exp_manager
   logger: False  # Provided by exp_manager
+  flush_logs_every_n_steps: 200
   log_every_n_steps: 100
-  check_val_every_n_epoch: 10
-  automatic_optimization: false
+  check_val_every_n_epoch: 1
```
Comment (blisc) [inline examples/tts/conf/hifigan/hifigan.yaml:57]: This seems far too frequent. Can we leave it at 10?

File: examples/tts/conf/hifigan/hifigan.yaml [hunk 5]
```diff
@@ -36,23 +36,31 @@ model:
...
+    warmup_ratio: 0.02
+    max_steps: ???
+
   denoise_strength: 0.0025
-  max_steps: 2500000
 
 trainer:
   gpus: -1 # number of gpus
-  max_steps: ${model.max_steps}
+  max_epochs: ???
```
Comment (blisc) [inline examples/tts/conf/hifigan/hifigan.yaml:49]: Why switch this?

File: examples/tts/conf/talknet-durs.yaml [hunk 6]
```diff
@@ -30,8 +30,9 @@ model:
         punct: true
         spaces: true
         stresses: false
+        add_blank_at: last
     dataloader_params:
-      drop_last: false
+      drop_last: true
```
Comment (blisc) [inline examples/tts/conf/talknet-durs.yaml:35]: This doesn't really do what you want in lightning. So I would advise you against adding this.

File: examples/tts/conf/talknet-durs.yaml [hunk 7]
```diff
@@ -148,6 +150,8 @@ model:
     sched:
       name: CosineAnnealing
       min_lr: 1e-5
+      warmup_ratio: 0.02
+      max_steps: ???
```
Comment (blisc) [inline examples/tts/conf/talknet-durs.yaml:154]: I don't think you need this. This should be computed by NeMo itself.

File: examples/tts/conf/talknet-pitch.yaml [hunk 8]
```diff
@@ -148,6 +154,8 @@ model:
     sched:
       name: CosineAnnealing
       min_lr: 1e-5
+      warmup_ratio: 0.02
+      max_steps: ???
```
Comment (blisc) [inline examples/tts/conf/talknet-pitch.yaml:158]: I don't think you need this. This should be computed by NeMo itself.

File: examples/tts/conf/talknet-spect.yaml [hunk 9]
```diff
@@ -215,6 +219,8 @@ model:
     sched:
       name: CosineAnnealing
       min_lr: 1e-5
+      warmup_ratio: 0.02
+      max_steps: ???
```
Comment (blisc) [inline examples/tts/conf/talknet-spect.yaml:223]: I don't think you need this. This should be computed by NeMo itself.

File: nemo/collections/tts/models/hifigan.py [hunk 10]
```diff
@@ -58,32 +57,41 @@ def __init__(self, cfg: DictConfig, trainer: 'Trainer' = None):
         self.discriminator_loss = DiscriminatorLoss()
         self.generator_loss = GeneratorLoss()
 
+        self.l1_factor = 45
+        self.num_log_audios = 5
```
Comment (blisc) [inline nemo/collections/tts/models/hifigan.py:61]: You should add this in the yaml instead of hard coding these.

File: nemo/collections/tts/models/hifigan.py [hunk 11]
```diff
@@ -58,32 +57,41 @@ def __init__(self, cfg: DictConfig, trainer: 'Trainer' = None):
...
         if self._train_dl and isinstance(self._train_dl.dataset, MelAudioDataset):
-            self.finetune = True
-            logging.info("fine-tuning on pre-computed mels")
+            self.input_as_mel = True
         else:
-            self.finetune = False
-            logging.info("training on ground-truth mels")
+            self.input_as_mel = False
+
+        self.automatic_optimization = False
```
Comment (blisc) [inline nemo/collections/tts/models/hifigan.py:71]: What is this flag?

File: nemo/collections/tts/models/hifigan.py [hunk 12]
```diff
@@ -58,32 +57,41 @@ def __init__(self, cfg: DictConfig, trainer: 'Trainer' = None):
...
         self.optim_d = instantiate(
             self._cfg.optim, params=itertools.chain(self.msd.parameters(), self.mpd.parameters()),
         )
 
-        max_steps = self._cfg.max_steps
-        warmup_steps = 0 if self.finetune else np.ceil(0.2 * max_steps)
         self.scheduler_g = CosineAnnealing(
-            self.optim_g, max_steps=max_steps, min_lr=1e-5, warmup_steps=warmup_steps
+            optimizer=self.optim_g,
+            max_steps=self._cfg.sched.max_steps,
```
Comment (blisc) [inline nemo/collections/tts/models/hifigan.py:81]: You should compute this instead of adding the user to provide this,

File: nemo/collections/tts/models/hifigan.py [hunk 13]
```diff
@@ -58,32 +57,41 @@ def __init__(self, cfg: DictConfig, trainer: 'Trainer' = None):
...
             'scheduler': self.scheduler_g,
             'interval': 'step',
         }
-        self.scheduler_d = CosineAnnealing(self.optim_d, max_steps=max_steps, min_lr=1e-5)
+
+        self.scheduler_d = CosineAnnealing(
+            optimizer=self.optim_d,
+            max_steps=self._cfg.sched.max_steps,
+            min_lr=self._cfg.sched.min_lr,
+        )
```
Comment (blisc) [inline nemo/collections/tts/models/hifigan.py:94]: You might want to consider adding warmup here as well

File: nemo/collections/tts/models/talknet.py [hunk 14]
```diff
@@ -136,11 +135,11 @@ def _metrics(self, true_f0, true_f0_mask, pred_f0_sil, pred_f0_body):
         sil_acc *= true_f0_mask.type_as(sil_acc)
         sil_acc = sil_acc.sum() / true_f0_mask.sum()
 
-        body_mse = F.mse_loss(pred_f0_body, (true_f0 - self.F0_MEAN) / self.F0_STD, reduction='none')
+        body_mse = F.mse_loss(pred_f0_body, (true_f0 - self.f0_mean) / self.f0_std, reduction='none')
```
Comment (blisc) [inline nemo/collections/tts/models/talknet.py:138]: is self.f0_mean and self.f0_std defined?

File: nemo/collections/tts/models/hifigan.py [hunk 15]
```diff
@@ -190,58 +199,58 @@ def validation_step(self, batch, batch_idx):
...
         if batch_idx == 0 and isinstance(self.logger, WandbLogger) and HAVE_WANDB:
             clips = []
             specs = []
-            for i in range(min(5, audio.shape[0])):
+            for i in range(min(self.num_log_audios, audio.shape[0])):
                 clips += [
                     wandb.Audio(
                         audio[i, : audio_len[i]].data.cpu().numpy(),
-                        caption=f"real audio {i}",
+                        caption=f"real audio {i}, epoch {self.current_epoch}",
```
Comment (blisc) [inline nemo/collections/tts/models/hifigan.py:217]: Does this produce multiple audio files on wandb? Or will it still show one audio file?

File: nemo/collections/tts/models/hifigan.py [hunk 16]
```diff
@@ -175,10 +183,11 @@ def training_step(self, batch, batch_idx, optimizer_idx):
             "global_step": self.global_step,
             "lr": self.optim_g.param_groups[0]['lr'],
         }
-        self.log_dict(metrics, on_step=False, on_epoch=True, sync_dist=True)
+        self.log_dict(metrics, on_step=True, sync_dist=True)
```
Comment (blisc) [inline nemo/collections/tts/models/hifigan.py:186]: on_step seems to be too frequent. Or will lightning automatically log at the log_every_n_steps interval?

File: nemo/collections/tts/models/hifigan.py [hunk 17]
```diff
@@ -146,20 +154,20 @@ def training_step(self, batch, batch_idx, optimizer_idx):
             disc_real_outputs=msd_score_real, disc_generated_outputs=msd_score_gen
         )
         loss_d = loss_disc_msd + loss_disc_mpd
-        self.manual_backward(loss_d, self.optim_d)
+        self.manual_backward(loss_d)
```
Comment (blisc) [inline nemo/collections/tts/models/hifigan.py:157]: Why did you remove the optimizer? Does this still work?

File: examples/tts/conf/hifigan/hifigan.yaml (hunk 4 above)
Comment (Oktai15) [reply examples/tts/conf/hifigan/hifigan.yaml:57]: +

File: examples/tts/conf/hifigan/hifigan.yaml (hunk 5 above)
Comment (Oktai15) [reply examples/tts/conf/hifigan/hifigan.yaml:49]: Because it is not valid code

File: examples/tts/conf/talknet-durs.yaml (hunk 6 above)
Comment (Oktai15) [reply examples/tts/conf/talknet-durs.yaml:35]: Ok.

File: examples/tts/conf/talknet-durs.yaml (hunk 7 above)
Comment (Oktai15) [reply examples/tts/conf/talknet-durs.yaml:154]: You mean I can remove `max_steps`?

File: nemo/collections/tts/models/hifigan.py (hunk 11 above)
Comment (Oktai15) [reply nemo/collections/tts/models/hifigan.py:71]: It allows to use loss.backward() manually, as we need it in GAN. It shoudn't be in config, because it is always for HiFiGAN and moreover previous flag didn't work (no valid code).

File: nemo/collections/tts/models/talknet.py (hunk 14 above)
Comment (Oktai15) [reply nemo/collections/tts/models/talknet.py:138]: Yes, in init.

File: nemo/collections/tts/models/hifigan.py (hunk 15 above)
Comment (Oktai15) [reply nemo/collections/tts/models/hifigan.py:215]: Still show one audio file.

File: nemo/collections/tts/models/hifigan.py (hunk 17 above)
Comment (Oktai15) [reply nemo/collections/tts/models/hifigan.py:155]: yes, this way was deprecated by PTL

File: nemo/collections/tts/models/hifigan.py (hunk 16 above)
Comment (Oktai15) [reply nemo/collections/tts/models/hifigan.py:184]: automatically

File: nemo/collections/tts/models/hifigan.py (hunk 10 above)
Comment (Oktai15) [reply nemo/collections/tts/models/hifigan.py:61]: Ok.

File: examples/tts/conf/hifigan/hifigan.yaml (hunk 5 above)
Comment (blisc) [reply examples/tts/conf/hifigan/hifigan.yaml:49]: What do you mean by it is not valid code? What was the error?

Comment (blisc) [reply examples/tts/conf/hifigan/hifigan.yaml:49]: I see, ok that is an error on my part. I will fix it

File: nemo/collections/tts/models/hifigan.py (hunk 17 above)
Comment (blisc) [reply nemo/collections/tts/models/hifigan.py:155]: It looks like if we do this, then we have to manually called our optimizer.step(). I would advise you to add in the .step() call

Comment (Oktai15) [reply nemo/collections/tts/models/hifigan.py:155]: I call opt.step() in the next line

File: nemo/collections/asr/data/vocabs.py (hunk 1 above)
Comment (stasbel) [reply nemo/collections/asr/data/vocabs.py]: i don't endorse this, but ok
better option would be to specify blank id inside CTC loss

File: examples/tts/conf/talknet-pitch.yaml (hunk 2 above)
Comment (stasbel) [reply examples/tts/conf/talknet-pitch.yaml:68]: could be random fluctuation
num_workers=0 does not spawn processes
num_workers=1 does spawn just single process, which does exactly the same thing

File: examples/tts/conf/talknet-pitch.yaml (hunk 3 above)
Comment (stasbel) [reply examples/tts/conf/talknet-pitch.yaml:15]: f0 statistics calculations not just fast, it's very very fast
we cache all the f0 values now, until this is happen we can calculate all the statistics immediately

File: nemo/collections/tts/models/hifigan.py (hunk 17 above)
Comment (blisc) [reply nemo/collections/tts/models/hifigan.py:155]: Oh good point, I missed that.

File: examples/tts/conf/hifigan/hifigan.yaml (hunk 5 above)
Comment (Oktai15) [reply examples/tts/conf/hifigan/hifigan.yaml:49]: @blisc it works!

**Review by blisc — APPROVED** [review]
LGTM, can merge so long as @stasbel has no objections with the talknet stuff

**Comment by stasbel** [comment]
LGTM
i will update ckpts links in following days, as well as TTS notebook for inference

**Comment by Oktai15** [comment]
@stasbel @blisc I found two additional bug in HiFiGAN tuning. I will fix it and after that we will merge it

**Comment by Oktai15** [comment]
@blisc let's merge it after this PR: https://github.com/NVIDIA/NeMo/pull/1994

File: examples/tts/conf/hifigan/model/validation_ds/val_ds_finetune.yaml [hunk 18]
```diff
@@ -3,9 +3,9 @@ dataset:
   _target_: "nemo.collections.tts.data.datalayers.MelAudioDataset"
   manifest_filepath: ${validation_datasets}
   min_duration: 3
-  n_segments: 66150
+  n_segments: 66048
```
Comment (blisc) [inline examples/tts/conf/hifigan/model/validation_ds/val_ds_finetune.yaml:6]: Why does validation use n_segments? Validation should just load the entire audio clip

Comment (Oktai15) [reply examples/tts/conf/hifigan/model/validation_ds/val_ds_finetune.yaml:6]: @blisc current implementation of `MelAudioDataset` requires this parameter and it should be divided by 256
